dcicutils/s3_utils.py

```diff
@@ -260,52 +271,73 @@ def health_page_get(self, health_page_key: str, *, default: Any = None, force: b
 
     ACCESS_KEYS_S3_KEY = 'access_key_admin'
 
-    def get_access_keys(self, name=ACCESS_KEYS_S3_KEY):
+    def get_access_keys(self, name: str = ACCESS_KEYS_S3_KEY) -> ServerAuthDict:
+        # TODO: Make .get_access_keys() do better error checking?
+        #       I have marked the return value as ServerAuthDict because it's CLEAR that various callers
+        #       assume that values coming back from this function are in the right form. e.g.,
+        #       ff_utils.get_authentication_with_server will give a hard error if this ever returns
+        #       something with a server missing.  So presumably we mostly do the right thing here.
+        #       It would be much better to properly error-check the result here or even in .get_key().
+        #       -kmp 1-Mar-2023
         keys = self.get_key(keyfile_name=name)
         if not isinstance(keys, dict):
             raise ValueError("Remotely stored access keys are not in the expected form")
 
         if isinstance(keys.get('default'), dict):
             keys = keys['default']
         if self.url:
+            # NOTE WELL: self.url might contain the empty string if S3Utils was initialized with a system bucket
+            #            instead of an ff_env. Moreover, we are overwriting whatever server was actually stored
+            #            on the server, so we might be overwriting good data with bad. We should document that in
+            #            fact self.url needs to always be set to a non-empty value by the time we reach here, and
+            #            probably raise an error otherwise. We should also raise an error if there's a server
+            #            in the data we retrieved because it's just not going to get used.  -kmp 1-Mar-2023
             keys['server'] = self.url
-        return keys
+        return keys  # Might still have no server if self.url was not a URL
 
-    def get_ff_key(self):
+    def get_ff_key(self) -> ServerAuthDict:
+        # TODO: The return value should be just AuthDict if we decide get_access_keys() can't guarantee ServerAuthDict.
         return self.get_access_keys()
 
-    def get_higlass_key(self):
+    def get_higlass_key(self) -> Union[AnyJsonData, str]:  # parsed json or string that won't parse
+        # TODO: It would be nice if this could be returning something with a better promise. -kmp 1-Mar-2023
```

Comment:
This actually will not return JSON data but raw text I believe. Same deal with Google/JH keys. I don't actually think the Google key is used anymore.

Reply:
I'm not sure this is still needed, but we can see about making a change another time. I added some notes on another comment a little farther down this same page.